rfd/0110-ip-pinning-and-propagation.md

```diff
@@ -0,0 +1,398 @@
+---
+authors: Anton Miniailo (anton@goteleport.com)
+state: implementation in progress
+---
+
+# RFD 0110 - TLS IP pinning and IP propagation.
+
+## Required Approvers
+
+- Engineering: `@r0mant` && `@strideynet` && `@codingllama`
+- Product: `@klizhentas || @xinding33`
+- Security: `@reedloden`
+
+
+## What
+
+We want to extend recently added SSH IP pinning functionality to TLS as well. It means having information about client's pinned IP stored on the TLS certificate and enforcing this check.
+
+## Why
+
+Current situation with only SSH certificates having IP pinning leaves the whole feature unfinished and easy to circumvent. Adding support for IP pinning to TLS certificates will add another layer of security which protects clients from their certificates being stolen by a malicious actor and used from another machine.
+
+## Details
+
+### Current situation
+Currently, IP pinning only works for SSH certificates, through generally supported critical option `source-address`. To complete IP pinning we'll need to implement similar functionality for TLS certificates.
+
+At the present, we have fields `clientIP` and `sourceIP` defined on `auth.certRequest`.
+```go
+type certRequest struct {
+	//...
+
+	// clientIP is an IP of the client requesting the certificate.
+	clientIP string
+	// sourceIP is an IP this certificate should be pinned to
+	sourceIP string
+}
+```
+
+For SSH certificates it in the end translates to extension field `ClientIP` and critical options `source-address` accordingly.
+But for TLS identity we currently only have field `ClientIP`, identity doesn't have analogue of `sourceIP`, related to specifically IP pinning functionality.
+
+### Fields changes
+
+For encoding pinning information, new TLS subject extension field will be added - `PinnedIP`, which will hold the client's pinned IP on certificate creation. Also, for consistency, field `SourceIP`, which exists on certificate requests (and currently does not exist on TLS identity), should be renamed to `PinnedIP` as well (name of the final SSH critical option `source-address` will remain the same, since it's a public established name). It will make intention of this field much more clear. Also, currently existing field `ClientIP` will be renamed to `LoginIP`.
+
+New struct will look like this:
+
+```go
+// Identity is an identity of the user or service, e.g. Proxy or Node
+type Identity struct {
+	//...
+
+	// LoginIP is an observed IP of the client on the moment of user certificate creation.
+	LoginIP string
+
+	// PinnedIP is an IP from which client is allowed to communicate with Teleport
+	PinnedIP string
+}
+
+// LoginIPASN1ExtensionOID is an extension ID used when encoding/decoding
+// the client login IP into certificates.
+LoginIPASN1ExtensionOID = asn1.ObjectIdentifier{1, 3, 9999, 1, 9}
+
+// PinnedIPASN1ExtensionOID is an extension ID used when encoding/decoding
+// the pinned IP into certificates.
+PinnedIPASN1ExtensionOID = asn1.ObjectIdentifier{1, 3, 9999, 2, 15}
```

Comment:
I'm a bit of a newbie to this feature, but this OID caught my eye. I haven't run across .1.3.9999 before, but it appears to be a reserved range of unassigned numbers.

My only worry with choosing these is that there is a potential for OID reuse or overlap in other organizations.

We can [request an IANA Private Enterprise Number](https://www.iana.org/assignments/enterprise-numbers/assignment/apply/) and get a globally-unique OID rooted at .1.3.6.1.4.1

Reply:
I don't think it should matter, because Teleport checks for the presence of these certificate extensions only if the certificate is valid and signed by Teleport's CA.

An external certificate with conflicting OIDs in the private range would not be trusted by Teleport and would fail authentication before we even got to inspecting the extensions.